### ISSUE TITLE
problem with putting GoogleServices-Info.plist

### ISSUE THREAD

**Issue by sonrebmax** [fechanique/cordova-plugin-fcm#141]
*** Terminating app due to uncaught exception 'com.firebase.core', reason: '[FIRApp configure] could not find a valid GoogleServices-Info.plist in your project. Please download one from https://console.firebase.google.com/.'

I added manually to:
www/
/

added scripts/after_prepare.js and hook in plugin.xml

no problem solved yet...

please help

**Comment by sonrebmax** [comment]
I have temporary solution for this:

cp GoogleService-Info.plist platforms/ios/build/device/APPNAME.app/ && ios-deploy --justlaunch --no-wifi -d -b platforms/ios/build/device/APPNAME.app

**Comment by GitFr33** [comment]
I am having the same problem.

**Comment by divinity-comp** [comment]
Same issue this is in phonegap for me 
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':processReleaseGoogleServices'.
> File google-services.json is missing. The Google Services Plugin cannot function without it. 
   Searched Location: 
  /project/src/release/google-services.json
  /project/google-services.json

but the file is directly in the root directory i evenattempted even /project/src/release creating that file structure

**Comment by BananZGan** [comment]
instead of putting them under the project's root directory, what i do is to put them under individual platform's directory.
For example:
I put the 'AndroidManifest.xml' file under *your_project*/platforms/android
and for ios, put 'GoogleService-Info.plist' under *your_project*/platforms/ios

**Comment by dannywillems** [comment]
@divinity-comp @sonrebmax @GitFr33: The solution is what @BananZGan said:
- For Android: put `google-services.json` in the directory `platforms/android`.
- For iOS: add the platform `ios`, and add `Google-Services.plist` in the root folder of the XCode project.

I hope we can have something like a hook which copies the files from the root Cordova project (so where `www` and `config.xml` are) automatically.

**Comment by fechanique** [comment]
I have some solutions for the hook, but I need to test them first

**Comment by dannywillems** [comment]
phonegap-plugin-push implements a simple hook. It could inspire.

**Comment by fechanique** [comment]
also I have some pull request with the solution, I only need some time :D

**Comment by GitFr33** [comment]
I'm still getting this error after copying my GoogleService-Info.plist into platforms/ios (and several other directories for good measure).

I'm using cordova version 6.3.1, Xcode 7.3.1, iOS 9.3

Any suggestions?

**Comment by benlooi** [comment]
I had this a few times. What I did was to add the GoogleService-Info.plist into the Resources folder in XCode. For some reason a blank GoogleService-Info.plist is created, which has to be replaced.
1) right-click the GoogleService-Info.plist under the Resources folder if it exists. If it shows up blank, delete it.
2) Next, right-click on Resources folder, add files, then find your downloaded plist and add it in. You can also try dragging it from your Finder window to the Resources folder in XCode.

**Comment by GitFr33** [comment]
Thanks benlooi! After adding GoogleService-Info.plist to the resources folder as you suggested it seems to be working. 
I am not getting any notifications unless the app is open, but that is a different issue.

**Comment by julitroalves** [comment]
Thanks @benlooi! After adding G..plist file to Resources my app finally works.

**Comment by arnoldmukisa** [comment]
Adding the G.plist to resources worked.

**Comment by madmars14** [comment]
Yes but this is a temporary solution as the issue persists in cordova 7.0.1 with xcode 8.3.2 as well . Has there any other fix been done ?

**Comment by DropkickSteve** [comment]
I think this is a bug of Xcode. When i add platform ios, the GoogleService-Info.plist is placed correctly in
/platforms/ios/myApp/Resources/. However it has to be manually added in the Resources folder of Xcodes project navigator.

**Comment by ricardojlrufino** [comment]
I think that in XCode Resources is a virtual folder... or is a Bug .. 

![image](https://user-images.githubusercontent.com/515675/26885126-1982838e-4b78-11e7-9417-9b4a67e34ba2.png)

**Comment by DropkickSteve** [comment]
@ricardojlrufino  I agree.
If I manually put the GoogleService-Info.plist file in the Resources folder in Xcode, it then appears in the filesystem in Finder. However when my script puts the file in the Resources folder of the filesystem, it does not show in the Resources folder of Xcode.

**Comment by manfield** [comment]
Hi Everybody, so there won't be any programmatic solution to that?

**Comment by bunday** [comment]
im still experiencing this error, about to remove and add ios platform and manually copy the file now. i hope it works

**Comment by manfield** [comment]
I found a programmatic solution, with after_prepare hook script

Adding these instructions:
```
  copyFile("../GoogleService-Info.plist", "../platforms/ios/GoogleService-Info.plist");
```

then

```
      var pbxGroupKey = proj.findPBXGroupKey({
        name: "Resources"
      });
      proj.removeResourceFile('GoogleService-Info.plist', {}, pbxGroupKey);
      proj.addResourceFile('GoogleService-Info.plist', {}, pbxGroupKey);
```

**Comment by atjason** [comment]
Just FYI, I solved this issue by add a copy step.
![2203101](https://user-images.githubusercontent.com/803199/29876364-29f060ee-8dcf-11e7-8e97-97e02fd28cfa.jpg)

The key point is, the 'GoogleService-Info.plist' should be copied to the root path of .app.
![2205161](https://user-images.githubusercontent.com/803199/29876434-57b7b54a-8dcf-11e7-9b45-00f19e60dbc4.jpg)

**Comment by manfield** [comment]
Ok, but since project is generated for each cordova build I think you'd get this setting lost after the next build.

**Comment by ank-it** [comment]
@atjason Thank you for pointing out the right solution. Works like charm. It should be in the official documentation of Firebase if this is not a bug.

**Comment by atjason** [comment]
@ank-it Welcome. I also wonder why Firebase official document miss such information.

**Comment by tumocpc** [comment]
I found the reason why the googleservice-info.plist file is corrupted.
This is plugin problem and I fixed it.
please add this plugin in your ionic project, and then it will fix the your problems.
https://github.com/tumocpc/cordova-plugin-fmc.git
It does not require work that you should add googleservice-info.plist manuallly.

**Comment by mkwsra** [comment]
In my case, I didn't download `GoogleService-info.plist` from firebase, thus, I was using always an empty version of it!
Plus, drag and drop this file into `Resources` folder didn't help me, I had to right click then choose `Add Files to FooBar`.

Thanks a lot for helping me out! I have been struggling with this issue!

**Comment by epetre** [comment]
@manfield has the best solution here. It's just missing some steps.

```
var projectPath = "platforms/ios/{YOUR_PROJECT_NAME}.xcodeproj/project.pbxproj";
var proj = new xcode.project(projectPath);
proj.parseSync();
var pbxGroupKey = proj.findPBXGroupKey({ name: "Resources" });
proj.removeResourceFile('GoogleService-Info.plist', {}, pbxGroupKey);
proj.addResourceFile('GoogleService-Info.plist', {}, pbxGroupKey);
fs.writeFileSync(projectPath, proj.writeSync());
```

_Just the parse and saving of the file._

I also copy my file from root, this way you can remove the platform and it still works:
```
fs.copyFile("GoogleService-Info.plist", "platforms/ios/{YOUR_PROJECT_NAME}/Resources/GoogleService-Info.plist", (err) => {
  if (err) throw err;
  console.log('GoogleService-Info.plist was copied to platforms/ios/GoogleService-Info.plist');
});
```

thx  @manfield

**Comment by rpcarnell** [comment]
For those of you having this problem and using build.phonegap.com:

In config.xml, inside <platform name="ios"> put   <resource-file src="resources/GoogleService-Info_2.plist" /> notice the _2. Change the name of the file to GoogleService-Info_2.plist and put it in the /resources file. 

The thing is build.phonegap, when it installs the FCM plugin creates its own GoogleService-Info.plist, deleting the original file. 

Rename the resulting IPA file. to *.zip. And open it, you will see both GoogleService-Info.plist and GoogleService-Info_2.plist. Remove GoogleService-Info.plist and replace it with GoogleService-Info_2.plist.
